### ISSUE TITLE
Allow Custom YAML Schema Validation Files

### ISSUE THREAD

**Issue by camball** [42Crunch/vscode-openapi#269]
## TL;DR

vscode-openapi forces the use of its own validation schema, overriding custom schemas.

## The Problem

In YAML files, when `openapi: 3.0.0` is specified, VS Code forces the use of the OpenAPI 3.0.X validation schema provided by vscode-openapi, and it doesn't appear there is any way to override it. I want to be able to upload my own custom schema file and not have vscode-openapi override it.

I've tried many different things to get the YAML extension (redhat.vscode-yaml) and vscode-openapi to play nicely, and I keep getting stuck because of the validation schema vscode-openapi applies by default. When I disable the extension, the problem goes away... but I'd love to be able to use both this extension for all its capabilities, as well as custom schemas! 😄

I've tried using glob patterns to try and ignore the default-applied validation schema, but this is not only a hacky solution, but also I couldn't get it to work:

```jsonc
// settings.json

{
    "yaml.schemas": {
        "schema/generated/openapi-3.0-2019-04-02.json": "!", // did not work, vscode-openapi schema still applied
        "v3.json": "!",                                      // did not work, vscode-openapi schema still applied
        "openapi:v3": "!",                                   // did not work, vscode-openapi schema still applied
    }
}
```

## A Potential Solution

I can see the code that causes the issue, where it registers the YAML contributor, `if (yamlExtension)`:

https://github.com/42Crunch/vscode-openapi/blob/07dcd43bbbb297b487409b813ded411f927381ce/src/yaml-schema-contributor.ts#L9-L13

https://github.com/42Crunch/vscode-openapi/blob/07dcd43bbbb297b487409b813ded411f927381ce/src/yaml-schema-contributor.ts#L56

I want to be able to upload my own custom schema file and not have vscode-openapi override it. I'd be happy to have a setting so that built-in validation through this extension can be disabled if necessary, like:

```jsonc
// settings.json

{
    // defaults to true so that out of the box, users have schema validation unless they need a custom schema
    "openapi.useDefaultValidationSchema": false,

     // could also be written something like the following
    "openapi.disableDefaultValidationSchema": false,
}
```

and this value is used to determine whether vscode-openapi will register a contributor or not:

```ts
// src/yaml-schema-contributor.ts

if (yamlExtension && shouldUseDefaultValidationSchema) { 
  provideYamlSchemas(context, cache, yamlExtension); 
} else { 
  // TODO log 
}
```

where `shouldUseDefaultValidationSchema` is the value read from `settings.json`.

**Comment by ak1394** [comment]
Hi @camball ! This seems reasonable. I'll revisit this issue once we release the next version of the extension, and I'll have a bit more time on my hands!

**Comment by ak1394** [comment]
Looks like I should be able to do a small release and include this request it it as well. I would like to clear one issue though, looking at the YAML extension it looks like it should use settings both from this extension, and also a user contributed ones?

I would like to clarify, why don't you want to apply OpenAPI schema + yours? What is your use case?

**Comment by camball** [comment]
Great—thanks!

Good thinking—it should apply the OpenAPI schema *+ mine*. My use case is for essentially "extended" schema validation. We have some extra keys we need to force a user to include in their schemas for our platform, and as part of our developer ecosystem, want to bring accessible schema validation to them via IDE support (IntelliJ + VS Code).

What would happen if a conflict occurs though, where a custom schema has the same key as one in the OpenAPI schema? I.e., how would resolutions be made? If one rule has to take precedence, I feel like the custom key should take precedence over the key from OpenAPI schema. Or, would two problems in VS Code be highlighted? My gut tells me one should take precedence, but let me know if you think otherwise. I don't anticipate any conflicts resulting from combining schemas for my use case, but just a thought; maybe it's a non-issue.

**Comment by ak1394** [comment]
I just come across this in the YAML extension docs: https://github.com/redhat-developer/vscode-yaml/wiki/Extension-API

In particular, the line that reads: 

`... it uses the schema provided by the contributor. It will also use any schemas associated via yaml.schemas that associate with the current file as well.`

Which got me thinking, if simply providing a setting to disable OpenAPI schemas is a wrong thing to do... 

As to your questions, I think your schema must be a complimentary to the OpenAPI one. For example, if OpenAPI schema says that `maxLength` property is integer, you're not going to say in your schema that it should be string?

And, if there are no conflicts between the two schemas then you should see errors (if any) related to OpenAPI validation and also errors related to your custom schema.

With this in mind, let me try to do apply some custom schema through settings.json on top of OpenAPI one and see if it works.

**Comment by ak1394** [comment]
Nope, it doesn't seem to combine extension provided schema with the one from the settings. I'm going to go ahead and publish the setting that allows/denies contribution of the OpenAPI schema as discussed. 

BTW, we have (in the paid offering) something called "Data Dictionaries" which allows to enforce formats of the properties in the schemas in a centralized manner, it works in the IDEs too: https://docs.42crunch.com/latest/content/concepts/data_dictionaries.htm maybe it is of some interest.

**Comment by camball** [comment]
That sounds great, thanks. Correct, our schema is complimentary to the OpenAPI one.

**Comment by ak1394** [comment]
I've published the new 4.25.2 version which introduces new `openapi.advanced.disableYamlSchemaContribution` setting.
Now, you should be able to take advantage of both OpenAPI and your custom one. Though, you'll likely have to create a schema that ties the two. You can take OpenAPI one from here `https://github.com/42Crunch/vscode-openapi/blob/master/schema/openapi-3.0-2019-04-02.json`

**Comment by camball** [comment]
This is great—thanks so much @ak1394 :smile:
